src/inmanta/ast/statements/__init__.py

```diff
@@ -149,54 +193,258 @@ def get_dataflow_node(self, graph: DataflowGraph) -> dataflow.NodeReference:
+    promise.
+    """
+
+    def __init__(self, static: StaticEagerPromise, responsible: DynamicStatement) -> None:
+        super().__init__()
+        self.static: StaticEagerPromise = static
+        self.responsible: DynamicStatement = responsible
+        self._waiter: Optional[Waiter] = None
+        self._promise: Optional[ProgressionPromise] = None
+
+    def set_waiter(self, waiter: Waiter) -> None:
+        self._waiter = waiter
+
+    def _acquire(self, variable: ResultVariable) -> None:
+        """
+        Entry point for the ResultVariable waiter: actually acquire the promise
+        """
+        if self._waiter is None:
+            # already fulfilled, no need to acquire progression promise anymore
+            return
+        assert self._promise is None
+        self._promise = variable.get_progression_promise(self.responsible)
+        self._waiter = None
+
+    def fulfill(self) -> None:
+        """
+        If a promise was already acquired, fulfills it, otherwise cancels the waiter so no new promise is acquired when the
+        variable becomes available.
+        """
+        if self._waiter is not None:
+            # waiter is still waiting, remove it
+            self._waiter.queue.remove_from_all(self._waiter)
+            self._waiter = None
+        if self._promise is not None:
+            self._promise.fulfill()
+
+    def variable_resume(
+        self,
+        variable: ResultVariable,
+        resolver: Resolver,
+        queue: QueueScheduler,
+    ) -> None:
+        self._acquire(variable)
+
+
 class ReferenceStatement(ExpressionStatement):
     """
     This class models statements that refer to other statements
     """
 
     def __init__(self, children: List[ExpressionStatement]) -> None:
         ExpressionStatement.__init__(self)
-        self.children = children
+        self.children: Sequence[ExpressionStatement] = children
         self.anchors.extend((anchor for e in self.children for anchor in e.get_anchors()))
 
     def normalize(self) -> None:
         for c in self.children:
             c.normalize()
 
+    def get_all_eager_promises(self) -> Iterator["StaticEagerPromise"]:
+        return chain(super().get_all_eager_promises(), *(subexpr.get_all_eager_promises() for subexpr in self.children))
+
     def requires(self) -> List[str]:
         return [req for v in self.children for req in v.requires()]
 
     def requires_emit(self, resolver: Resolver, queue: QueueScheduler) -> Dict[object, ResultVariable]:
-        return {rk: rv for i in self.children for (rk, rv) in i.requires_emit(resolver, queue).items()}
+        parent_req: Mapping[object, ResultVariable] = super().requires_emit(resolver, queue)
+        own_req: Mapping[object, ResultVariable] = {
+            rk: rv for i in self.children for (rk, rv) in i.requires_emit(resolver, queue).items()
+        }
+        return {**parent_req, **own_req}
```

Comment:
This is bang on the critical path, I don't know if this affect performance, but it might.

Reply:
Seemingly negligible performance impact (still double checking), so I'm leaving this as is unless you have objections.